scripts/datalad_crawlers/ftp_crawler.py

```diff
@@ -2,24 +2,37 @@
 import os
 from ftplib import FTP
 
+import git
+from tqdm import tqdm
 
-def crawl(host, root, subdir):
+
+def crawl(host, root, subdir, *, ftp):
     """
     Recursively prints the URLs of the files under dir and adds them to git-annex
     """
-    old_dir = ftp.pwd()
-    ftp.cwd(os.path.join("/", root, subdir))
-    for entry in ftp.mlsd():
-        file_name, metadata = entry
-        if metadata["type"] == "dir":
-            print(f"mkdir -p {file_name}")
-            crawl(host, root, os.path.join(subdir, file_name))
-        elif metadata["type"] == "file":
-            url = f"ftp://{host}/" + os.path.join(root, subdir, file_name)
-            local_file_name = os.path.join(subdir, file_name)
-            print(f"git-annex addurl {url} --file {local_file_name}")
-            os.system(f"git-annex addurl {url} --file {local_file_name}")
-    ftp.cwd(old_dir)
+    cwd = os.path.join("/", root, subdir)
+    print(f"\nCrawling {cwd}")
+
+    for filename, metadata in tqdm(ftp.mlsd(cwd)):
+        filepath = os.path.join(subdir, filename)
+
+        while True:
+            try:
+                if metadata.get("type") == "dir":
+                    crawl(host, root, filepath, ftp=ftp)
+
+                elif metadata.get("type") == "file":
+                    git.Repo().git.annex(
+                        "addurl",
+                        f"ftp://{host}/" + os.path.join(cwd, filename),
+                        file=filepath,
+                    )
+                break
+
+            except Exception as e:
+                print(f"WARNING: Connection restarted on file: {filepath}: {e}")
```

Comment:
I see that the FTP connection is reset in here but if the connection failed on the file, doesn't it mean that this file is skipped and it goes to the next filename?

Reply:
@mathdugre did you answer my question above? Sorry if I don't remember. I don't see anything in the comments of the PR as far as I can tell.